stan/math/rev/mat/functor/cvodes_integrator.hpp

```diff
@@ -0,0 +1,376 @@
+        // N_VMake_Serial sets own_data to false
+        N_VDestroy_Serial(cvode_state_);
+        if (cvode_state_sens_ != NULL)
+          N_VDestroyVectorArray_Serial(cvode_state_sens_, S_);
+        CVodeFree(&cvode_mem_);
+      }
+
+      /**
+       * Forward the CVode array-based call from
+       * the ODE RHS to the Stan vector-based call
+       *
+       * @param[in] y state of the base ODE system
+       * @param[out] dy_dt ODE RHS at time t
+       * @param[in] t time
+       */
+      void rhs(const double y[], double dy_dt[], double t) const {
+        const std::vector<double> y_vec(y, y + N_);
+
+        std::vector<double> dy_dt_vec;
+        ode_model_(y_vec, dy_dt_vec, t);
+
+        std::copy(dy_dt_vec.begin(), dy_dt_vec.end(), dy_dt);
+      }
+
+      // Static wrapper for CVode callback
+      static int ode_rhs(double t, N_Vector y, N_Vector ydot, void* f_data) {
+        ode const* explicit_ode = reinterpret_cast<ode const*>(f_data);
+        explicit_ode->rhs(NV_DATA_S(y), NV_DATA_S(ydot), t);
+        return 0;
+      }
+
+      /**
+       * Calculate the sensitivity RHS for the requested variables
+       * (initials and/or parameters). Function signature is
+       * pre-defined by CVODES library.
+       *
+       * @param[in] M number of parameters for which sensitivites are calculated
+       * @param[in] t time
+       * @param[in] y state of the base ODE system
+       * @param[in] ydot state of the RHS of the ODE system
+       * @param[in] yS array of M N_Vectors of size N, i.e. state of sensitivity
+       * RHS
+       * @param[out] ySdot array of M N_Vectors of size N of the sensitivity RHS
+       */
+      void rhs_sens(int M, realtype t,
+                    double y[], double ydot[],
+                    N_Vector *yS, N_Vector *ySdot) const {
+        const std::vector<double> y_vec(y, y + N_);
+
+        Eigen::VectorXd fy(N_);
+        Eigen::MatrixXd Jy(N_, N_);
+
+        if (param_var_) {
+          Eigen::MatrixXd Jtheta(N_, M_);
+          ode_model_.jacobian_SP(t, y_vec, fy, Jy, Jtheta);
+
+          for (size_t m = 0; m < M_; m++) {
+            // map NV_Vector to Eigen facilities
+            Eigen::Map<Eigen::VectorXd>
+              yS_eig(NV_DATA_S(yS[param_var_ind_ + m]), N_);
+            Eigen::Map<Eigen::VectorXd>
+              ySdot_eig(NV_DATA_S(ySdot[param_var_ind_ + m]), N_);
+
+            ySdot_eig = Jy * yS_eig + Jtheta.col(m);
+          }
+        } else {
+          ode_model_.jacobian_S(t, y_vec, fy, Jy);
+        }
+
+        if (initial_var_) {
+          for (size_t m = 0; m < N_; m++) {
+            // map NV_Vector to Eigen facilities
+            Eigen::Map<Eigen::VectorXd> yS_eig(NV_DATA_S(yS[m]), N_);
+            Eigen::Map<Eigen::VectorXd> ySdot_eig(NV_DATA_S(ySdot[m]), N_);
+
+            ySdot_eig = Jy * yS_eig;
+          }
+        }
+      }
+
+      // Static wrapper for CVode callback
+      static int ode_rhs_sens(int Ns, realtype t,
```

Comment:
Sorry --- hadn't read far enough.  Yes, that's what I meant.
The global means it doesn't need the class instance.

> On Apr 20, 2016, at 6:20 PM, Michael Betancourt <notifications@github.com> wrote:
> 
> In stan/math/rev/mat/functor/cvodes_integrator.hpp:
> 
> > +          ode_model_.jacobian_S(t, y_vec, fy, Jy);
> > +        }
> > +
> > +        if (initial_var_) {
> > +          for (size_t m = 0; m < N_; m++) {
> > +            // map NV_Vector to Eigen facilities
> > +            Eigen::Map<Eigen::VectorXd> yS_eig(NV_DATA_S(yS[m]), N_);
> > +            Eigen::Map<Eigen::VectorXd> ySdot_eig(NV_DATA_S(ySdot[m]), N_);
> > +
> > +            ySdot_eig = Jy * yS_eig;
> > +          }
> > +        }
> > +      }
> > +
> > +      // Static wrapper for CVode callback
> > +      static int ode_rhs_sens(int Ns, realtype t,
> 
> No. The static is needed to ensure that the method can be accessed with a global function pointer, which then allows us to pass it to the C callbacks in CVODE(S).
> …
> On Apr 20, 2016, at 11:18 PM, Daniel Lee ***@***.***> wrote: In stan/math/rev/mat/functor/cvodes_integrator.hpp: > + ode_model_.jacobian_S(t, y_vec, fy, Jy); > + } > + > + if (initial_var_) { > + for (size_t m = 0; m < N_; m++) { > + // map NV_Vector to Eigen facilities > + Eigen::Map<Eigen::VectorXd> yS_eig(NV_DATA_S(yS[m]), N_); > + Eigen::Map<Eigen::VectorXd> ySdot_eig(NV_DATA_S(ySdot[m]), N_); > + > + ySdot_eig = Jy * yS_eig; > + } > + } > + } > + > + // Static wrapper for CVode callback > + static int ode_rhs_sens(int Ns, realtype t, I think I see why this happened. This was to initialize it in the constructor? If it was done after the constructor, I'm guessing we don't need the static function? … — You are receiving this because you were mentioned. Reply to this email directly or view it on GitHub
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
>